### PR TITLE
code review cleanups

### PR DESCRIPTION
**Pull request by eeue56**
Various code review cleanups from Marica's reviews. 

Discovered some design patterns I want to fix outside of this PR:

1) functions and statements should return something like `ChannelMessage`. Basically, no user-defined function should really be called `self.send_channel_message` - instead, it should just return things and whatever it returns gets sent. This will help reduce multiple payloads and make point 2 easier to implement
2) All functions should return things to the channel for the channel to use (e.g `NOW` should send the current time to the slack channel)
3) It should be way easier to manipulate the main parse loop. I've done this before via `register_step : (a -> M b) -> M a -> M b`, basically a function that is injected into the main parse loop and can manipulate

### CODE REVIEW

**Comment by ento** [comment]
assigned based on the WIP tag in the title

File: bot_file.py [hunk 1]
```diff
@@ -182,7 +183,7 @@ def stop_listening(self, channel: str, pattern: str) -> None:
         self.notify.forget_pattern(person, pattern)
         self.notify.save_to_file(self.notify_file)
 
-    def time_to_show(self, channel: str, person: str) -> None:
+    def ping_person(self, channel: str, person: str) -> None:
         """ notify a person about a message """
         if channel.startswith('D'):
```
Comment (ento) [inline bot_file.py]: What's a channel name that starts with D? DMs? ..looks like yeah, although I couldn't find a documentation that explicitly states that. (closest is various encoded IDs mentioned in https://api.slack.com/methods/chat.postMessage)

:art: Maybe add a comment that we don't support IM channels (for now?)

File: bot_file.py [hunk 2]
```diff
@@ -244,8 +245,11 @@ def find_elm_017_matches(self, channel: str, filename_pattern: str) -> None:
 
         self.send_channel_message(channel, message)
 
-    def how_trivial_to_port(self, channel: str, filename_pattern: str) -> None:
-        """ give a filename of elm to get me to tell you how trivial it is to port """
+    def how_hard_to_port(self, channel: str, filename_pattern: str) -> None:
+        """ give a filename of elm to get me to tell you how trivial it is to port
```
Comment (ento) [inline bot_file.py]: Comment still says "trivial".

File: our_repo.py [hunk 3]
```diff
@@ -13,20 +15,20 @@ def __init__(self, folder: str, token: str, org: str, repo: str):
...
         os.makedirs(self.folder, exist_ok=True)
 
-    def _git_init(self):
+    def _git_init(self) -> None:
         os.system(f'git clone --depth 1 https://{self.token}@github.com/{self.org}/{self.repo}.git')
         os.chdir(self.repo)
         os.system(f'git remote set-url origin https://{self.token}@github.com/{self.org}/{self.repo}.git')
 
-    def _git_clone(self, branch_name='master'):
+    def _git_clone(self, branch_name='master') -> None:
```
Comment (ento) [inline our_repo.py]: `branch_name: str = 'master'`

File: our_repo.py [hunk 4]
```diff
@@ -13,20 +15,20 @@ def __init__(self, folder: str, token: str, org: str, repo: str):
...
         os.system(f'git remote set-branches origin {branch_name}')
         os.system(f'git fetch --depth 1 origin {branch_name}')
         os.system(f'git checkout origin/{branch_name}')
 
-    def get_ready(self, branch_name='master'):
+    def get_ready(self, branch_name='master') -> None:
```
Comment (ento) [inline our_repo.py]: `branch_name: str = 'master'`

File: our_repo.py [hunk 5]
```diff
@@ -80,24 +83,24 @@ def number_of_016_files(self):
...
         return elm_016_count
 
-    def get_files_for_017(self, pattern):
+    def get_files_for_017(self, pattern: str) -> List[str]:
         pattern = pattern.replace('.', '/')
         all_files = glob.glob(f'{self.repo_dir}/**/{pattern}.elm', recursive=True)
 
-        return [ file for file in all_files
-            if self.what_kinda_file(file) == ELM_017_FILE
+        return [ filename for filename in all_files
```
Comment (ento) [inline our_repo.py]: 😻

File: our_repo.py [hunk 6]
```diff
@@ -80,24 +83,24 @@ def number_of_016_files(self):
...
+            if self.what_kinda_file(filename) == ELM_017_FILE
         ]
 
-    def get_files_for_017_with_breakdown(self, pattern):
+    def get_files_for_017_with_breakdown(self, pattern: str) -> Dict[str,Dict[str, int]]:
```
Comment (ento) [inline our_repo.py]: 💭 should be `get_files_for_016_with_breakdown`? or

- `get_016_files_with_breakdown`
- `get_017_porting_breakdown`

File: our_repo.py [hunk 7]
```diff
@@ -122,7 +125,7 @@ def how_trivial_to_port(self, filename):
 
 
 
-    def what_kinda_file(self, filename):
+    def what_kinda_file(self, filename: str) -> int:
```
Comment (ento) [inline our_repo.py]: :art: `int` is scary - `WhatKindaFile` (enum?) would be nicer. or `ElmVersion`

File: our_repo.py [hunk 8]
```diff
@@ -80,24 +83,24 @@ def number_of_016_files(self):
...
 
-        return { file:self.how_trivial_to_port(file) for file in all_files
-            if self.what_kinda_file(file) == ELM_016_FILE
+        return { filename : self.how_trivial_to_port(filename) for filename in all_files
+            if self.what_kinda_file(filename) == ELM_016_FILE
         }
 
 
-    def how_trivial_to_port(self, filename):
+    def how_trivial_to_port(self, filename: str) -> Dict[str, int]:
```
Comment (ento) [inline our_repo.py]: this going to stay trivial?

File: notify.py [hunk 9]
```diff
@@ -4,25 +4,36 @@
...
 
-    def add_pattern(self, person: str, pattern: str):
+    def add_pattern(self, person: str, pattern: str) -> None:
+        """ register a pattern to notify a given person
+        """
+        
         if person not in self.patterns:
             self.patterns[person] = []
 
         self.patterns[person].append(pattern)
```
Comment (ento) [inline notify.py]: :art: could do a `re.compile(pattern)` here and notify the user if it fails, even without a message example to match (can be outside this PR)

File: bot_file.py [hunk 10]
```diff
@@ -49,15 +49,16 @@ def __init__(self, *args, **kwargs):
...
 
-        if who_wants_it is not None:
-            for person in who_wants_it:
-                self.time_to_show(message['channel'], person)
+        if people_who_want_notification is not None:
+            for person in people_who_want_notification:
+                self.ping_person(message['channel'], person)
+
 
     def parse_direct_message(self, message):
```
Comment (ento) [inline bot_file.py]: noticed that `now_statement` has the same doc comment as `wait_statement` way down↓↓↓

File: our_repo.py [hunk 11]
```diff
@@ -5,6 +5,8 @@
 
 import os
 import glob
+from typing import List, Tuple, Dict
```
Comment (ento) [inline our_repo.py]: :art: don't see `Tuple` used, but happy to defer those cleanups to after flake8

File: bot_file.py (hunk 1 above)
Comment (eeue56) [inline bot_file.py]: Added a comment to this method, and made it use `is_direct_message` instead

File: bot_file.py (hunk 10 above)
Comment (eeue56) [inline bot_file.py]: :+1:

File: bot_file.py (hunk 2 above)
Comment (eeue56) [inline bot_file.py]: fixed

File: our_repo.py (hunk 7 above)
Comment (eeue56) [inline our_repo.py]: ElmVersion enum

**Comment by eeue56** [comment]
sat8oct04 54 38bst2016

**Comment by eeue56** [comment]
That's better! though I think it still needs some extra tweaking - I'm not if things like `.` or `.*` should be allowed without other patterns to restrict it

**Comment by eeue56** [comment]
@ento back to you!

File: our_repo.py (hunk 11 above)
Comment (ento) [inline our_repo.py]: 👍

File: bot_file.py [hunk 12]
```diff
@@ -172,8 +174,15 @@ def when_you_hear(self, channel: str, pattern: str) -> None:
         """ notify the user when you see a pattern """
         person = self._last_sender
 
+        try:
+            re.compile(pattern)
```
Comment (ento) [inline bot_file.py]: re: detecting `.` / `.*`

```python
>>> import sre_parse
>>> sre_parse.parse('..*.{0,}')
[('any', None), ('max_repeat', (0, 4294967295, [('any', None)])), ('max_repeat', (0, 4294967295, [('any', None)]))]
```

could be useful. hmm, or keep a message history of 1000 or so and refuse to register a pattern that's going to be too noisy?

and/or deliver pings to the `@slackbot` channel.

File: bot_file.py [hunk 13]
```diff
@@ -182,9 +191,9 @@ def stop_listening(self, channel: str, pattern: str) -> None:
         self.notify.forget_pattern(person, pattern)
         self.notify.save_to_file(self.notify_file)
 
-    def time_to_show(self, channel: str, person: str) -> None:
-        """ notify a person about a message """
-        if channel.startswith('D'):
+    def ping_person(self, channel: str, person: str) -> None:
+        """ notify a person about a message. Ignore direct messages """
```
Comment (eeue56) [inline bot_file.py]: Will come back to this in another PR - you're right, it should probably be scoped to the right channel

**Pull request merged**